Clippy lint fixes

jb55 opened the pull request:
Continues from #24 so review that first.

Ones that confused me the most:

```
lint: remove noop drops
lint: silence dubious clippy warning about unsafe blocks
```

perhaps those indicate real issues but I'm not that familiar with unsafe block symantics

jb55 committed:
lint: clean up redundant static lint warnings

warning: Constants have by default a `'static` lifetime
  --> src/errors.rs:17:34
   |
17 | pub const GDK_ERROR_ID_UNKNOWN: &'static str = "id_unknown";
   |                                 -^^^^^^^---- help: consider removing `'static`: `&str`
   |
   = note: #[warn(clippy::redundant_static_lifetimes)] on by default

jb55 committed:
lint: cleanup unneeded returns

warning: unneeded return statement
   --> src/wallet.rs:158:21
    |
158 |             None => return Err(Error::WalletNotRegistered),
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove `return`: `Err(Error::WalletNotRegistered)`
    |
    = note: #[warn(clippy::needless_return)] on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return

jb55 committed:
lint: remove dead code in GA_auth_handler

We will most likely be removing all this anyways

warning: method is never used: `error`
  --> src/lib.rs:88:5
   |
88 |     fn error(err: String) -> *const GA_auth_handler {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

jb55 committed:
lint: fix clippy::inter_iter_on_ref issues

warning: this .into_iter() call is equivalent to .iter() and will not move the Vec
   --> src/coins/liq.rs:148:82
    |
148 |         let mut details = prevtx.as_object().req()?["details"].as_array().req()?.into_iter();
    |                                                                                  ^^^^^^^^^ help: call directly: `iter`
    |
    = note: #[warn(clippy::into_iter_on_ref)] on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#into_iter_on_ref

jb55 committed:
lint: fix clippy::cast_lossless issues

warning: casting u32 to u64 may become silently lossy if you later change the type
   --> src/coins/liq.rs:149:72
    |
149 |         let detail = match details.find(|d| d["vout"].as_u64() == Some(prevout.vout as u64)) {
    |                                                                        ^^^^^^^^^^^^^^^^^^^ help: try: `u64::from(prevout.vout)`
    |
    = note: #[warn(clippy::cast_lossless)] on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cast_lossless

jb55 committed:
lint: prefer destructuring to match

warning: you seem to be trying to use match for destructuring a single pattern. Consider using `if let`
   --> src/errors.rs:87:9
    |
87  | /         match e {
88  | |             bitcoincore_rpc::Error::JsonRpc(ref e) => match e {
89  | |                 jsonrpc::Error::Rpc(ref e) => match e.code {
90  | |                     CORE_INSUFFICIENT_FUNDS => return Error::InsufficientFunds,
...   |
100 | |             _ => {}
101 | |         }
    | |_________^
    |
    = note: #[warn(clippy::single_match)] on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_match
help: try this
    |
87  |         if let bitcoincore_rpc::Error::JsonRpc(ref e) = e { match e {
88  |     jsonrpc::Error::Rpc(ref e) => match e.code {
89  |         CORE_INSUFFICIENT_FUNDS => return Error::InsufficientFunds,
90  |         CORE_WALLET_GENERIC => {
91  |             if e.message.contains("Duplicate -wallet filename specified.") {
92  |                 return Error::AlreadyLoggedIn;

jb55 committed:
lint: convert &String to &str where possible

warning: writing `&String` instead of `&str` involves a new object where a slice will do.
   --> src/network.rs:196:22
    |
196 | fn read_cookie(path: &String) -> Result<(String, String), Error> {
    |                      ^^^^^^^ help: change this to: `&str`
    |
    = note: #[warn(clippy::ptr_arg)] on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg

jb55 committed:
lint: fix single-character string constants

warning: single-character string constant used as pattern
   --> src/network.rs:198:43
    |
198 |     let parts: Vec<&str> = contents.split(":").collect();
    |                                           ^^^ help: try using a char instead: `':'`
    |
    = note: #[warn(clippy::single_char_pattern)] on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_char_pattern

jb55 committed:
lint: allow error wildcard matching where it makes sense

warning: Err(_) will match all errors, maybe not a good idea
   --> src/wallet.rs:162:21
    |
162 |                     Err(_) => panic!(
    |                     ^^^^^^
    |
    = note: #[warn(clippy::match_wild_err_arm)] on by default
    = note: to remove this warning, match each error separately or use unreachable macro
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#match_wild_err_arm

jb55 committed:
lint: pass values with storage size <= 8 by value

No point in passing these by reference

jb55 committed:
lint: silence dubious clippy warning about unsafe blocks

Not sure if this is legit or not

error: this public function dereferences a raw pointer but is not marked `unsafe`
  --> src/session.rs:92:23
   |
92 |         Ok(unsafe { &*sess })
   |                       ^^^^
   |
   = note: #[deny(clippy::not_unsafe_ptr_arg_deref)] on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#not_unsafe_ptr_arg_deref

jb55 committed:
lint: equality checks against false can be replaced by a negation

warning: equality checks against false can be replaced by a negation
   --> src/wallet.rs:465:12
    |
465 |         if accept["allowed"].as_bool().req()? == false {
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try simplifying it as shown: `!(accept["allowed"].as_bool().req()?)`
    |
    = note: #[warn(clippy::bool_comparison)] on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#bool_comparison

jb55 committed:
lint: remove identical conversions

warning: identical conversion
   --> src/wallet.rs:573:41
    |
573 |           let mut estimates: Vec<Value> = (2u16..25u16)
    |  _________________________________________^
574 | |             .into_iter()
    | |________________________^ help: consider removing `.into_iter()`: `(2u16..25u16)`
    |
    = note: #[warn(clippy::identity_conversion)] on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#identity_conversion

jb55 committed:
lint: use if else instead of matching on a boolean

warning: you seem to be trying to match on a boolean expression
   --> src/wally.rs:233:17
    |
233 |       let flags = match segwit {
    |  _________________^
234 | |         false => 0,
235 | |         true => ffi::WALLY_TX_FLAG_USE_WITNESS,
236 | |     };
    | |_____^ help: consider using an if/else expression: `if segwit { ffi::WALLY_TX_FLAG_USE_WITNESS } else { 0 }`
    |
    = note: #[warn(clippy::match_bool)] on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#match_bool

jb55 committed:
lint: use ! .is_empty() instead of .len() > 0

warning: length comparison to zero
   --> src/lib.rs:270:8
    |
270 |     if read_str(password).len() > 0 {
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: using `is_empty` is clearer and more explicit: `!read_str(password).is_empty()`
    |
    = note: #[warn(clippy::len_zero)] on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#len_zero

stevenroose commented on src/wallet.rs:
```diff
@@ -219,12 +219,12 @@ impl Wallet {
-            internal_xpriv: internal_xpriv,
+            master_xpriv,
+            external_xpriv,
+            internal_xpriv,
```
Hmm, I don't really like this "optimization".. It's clearly convoluting a variable name and a field name which are two different concepts that just happen to share the same name.

Could we `#[allow(clippy::redundant_field_names)]` perhaps?

stevenroose commented on src/network.rs:
```diff
@@ -161,7 +161,7 @@ impl Network {
     }
 
-    pub fn get(id: &String) -> Option<&'static Network> {
+    pub fn get(id: &str) -> Option<&'static Network> {
```
Hmm, I remember doing this and feeling annoyed by it. Something complained about `&str` when I did this. If that seems to be solved, great! :)

stevenroose commented on src/session.rs:
```diff
@@ -103,7 +103,6 @@ impl SessionManager {
         if !self.sessions.remove(&sess) {
             throw!("session is unmanaged");
         }
-        unsafe { drop(&*sess) };
```
Hmm, are you sure no drop of some kind is required here?

jb55 replied:
yeah this is the one I was dubious about, the linter claims this is a no-op. I don't know enough about rust FFI / memory management to comment on this yet. I'll drop the commit for now and keep the lint warning/error

stevenroose commented on src/wally.rs:
```diff
@@ -230,9 +230,10 @@ pub fn tx_get_elements_signature_hash(
     sighash: u32,
     segwit: bool,
 ) -> sha256d::Hash {
-    let flags = match segwit {
```
Heh, yeah I do that because it's a line shorter and the logic is trivial. Anyhow, this is ok.

jb55 committed:
lint: allow redunant field names in wallet.rs

jb55 commented:
v2: git range-diff 0b9b27b...654708b updates:

- removed e14b099 `lint: cleanup redunant field names in struct initializations`
- removed a56e192 `lint: remove noop drops`
- added 654708b `lint: allow redunant field names in wallet.rs`

stevenroose commented:
reACK 654708b
